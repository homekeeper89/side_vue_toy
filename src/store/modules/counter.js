import {
  SET_COUNTER,
  GET_COUNTER,
  COUNTER2_M,
} from '@/store/modules/counter-types';

export const counter = {
  namespaced: true,
  state: {
    counter: 0,
  },

  actions: {
    [SET_COUNTER]: ({ commit, dispatch }) => {
      commit(SET_COUNTER);
      dispatch(`${COUNTER2_M}${SET_COUNTER}`, null, { root: true });
    },
  },
  mutations: {
    [SET_COUNTER]: (state) => state.counter++,
  },
  getters: {
    [GET_COUNTER]: (state) => state.counter,
  },
};
