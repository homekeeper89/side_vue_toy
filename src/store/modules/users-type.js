export const USER_NAMESPACE = 'users';

export const REGISTER_USER = 'REGISTER_USER';
export const CHECK_EMAIL = 'CHECK_EMAIL';
export const SET_USER_API_STATUS = 'SET_USER_API_STATUS';
