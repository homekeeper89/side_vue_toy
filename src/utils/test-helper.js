import MockAdapter from 'axios-mock-adapter';
//https://stackoverflow.com/questions/55038758/how-to-work-with-path-parameters-in-axios-mock-adapter
export class MockAxiosHelper {
  constructor(axios) {
    this.mockAxios = new MockAdapter(axios);
  }

  getAxios() {
    return this.mockAxios;
  }

  getUrl(methods) {
    return this.mockAxios.handlers[methods][0][0];
  }

  getData(methods) {
    return this.mockAxios.history[methods][0].data;
  }
}
