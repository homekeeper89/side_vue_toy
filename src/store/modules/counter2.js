import { SET_COUNTER, GET_COUNTER2 } from '@/store/modules/counter-types';

export const counter2 = {
  namespaced: true,
  state: {
    counter2: 0,
  },
  actions: {
    [SET_COUNTER]: ({ commit }) => commit(SET_COUNTER),
  },
  mutations: {
    [SET_COUNTER]: (state) => state.counter2++,
  },
  getters: {
    [GET_COUNTER2]: (state) => state.counter2,
  },
};
