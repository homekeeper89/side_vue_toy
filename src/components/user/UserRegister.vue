<template>
  <div class="UserRegister">
    <div>
      <span>email : </span>
      <input class="email" type="text" v-model="data.email" />
    </div>
    <div>
      <span>password : </span>
      <input class="password" type="text" v-model="data.password" />
    </div>
    <div>
      <input class="nickName" type="text" v-model="data.nickName" />
    </div>
    <div>
      <button class="userRegister" @click="register">등록</button>
    </div>
  </div>
</template>
<script>
import { mapActions } from 'vuex';
import { USER_NAMESPACE, REGISTER_USER } from '@/store/modules/users-type';
export default {
  name: 'UserRegister',
  data() {
    return {
      data: {
        email: '',
        password: '',
        nickName: '',
      },
    };
  },
  methods: {
    ...mapActions({ registerUser: `${USER_NAMESPACE}/${REGISTER_USER}` }),
    register() {
      let payload = this.data;
      this.registerUser(payload);
    },
    validateData() {
      let res = Object.keys(this.data).filter((item) => !this.data[item]);
      return res.length != 0 ? false : true;
    },
  },
};
</script>
