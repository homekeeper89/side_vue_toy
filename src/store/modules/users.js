import axios from 'axios';
import { REGISTER_USER, SET_USER_API_STATUS } from './users-type.js';
import { inspectResponse } from '@/utils/apiHelper';

const api_register_user = '/api/users/v1';

const reqresApi = axios.create({
  baseURL: 'https://reqres.in', // Url
  timeout: 5000, // timeout 5ì´ˆ
});

export const userStore = {
  namespaced: true,
  state: {
    apiStatus: {
      code: 0,
      msg: '',
    },
  },
  getters: {},
  actions: {
    [REGISTER_USER]: async ({ commit }, data) => {
      try {
        let res = await axios.post(api_register_user, {
          data: data,
        });
        inspectResponse(res);
        commit(SET_USER_API_STATUS, res);
      } catch (err) {
        console.error(err);
        throw Error('API Error occurred');
      }
    },
    testActions: async ({ commit }, data) => {
      try {
        let res = await axios.post(api_register_user, { data });
        inspectResponse(res.data);
        commit(SET_USER_API_STATUS, res.data);
      } catch (err) {
        console.error(err);
        throw Error('API Error occurred');
      }
    },
  },
  mutations: {
    [SET_USER_API_STATUS]: (state, payload) => {
      state.apiStatus.code = payload.status_code;
      state.apiStatus.msg = payload.msg;
    },
  },
};
