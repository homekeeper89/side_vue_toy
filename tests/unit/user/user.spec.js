import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import UserRegister from '@/components/user/UserRegister.vue';
import flushPromises from 'flush-promises';
import { users } from '@/store/modules/users';
import mockAxios from 'axios';
import { getUrlFromSpy, getDataFromSpy } from '../../test-helper.js';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('User Register 와 관련된 테스트', () => {
  const status = {
    status_code: 200,
    msg: 'success',
  };
  const email = 'test@username';
  const password = 'testPassword';
  const nickname = 'testnickname';
  let wp;
  let store;
  let spyPost;
  beforeEach(() => {
    mockAxios.post.mockImplementationOnce(() =>
      Promise.resolve({ data: status })
    );
    spyPost = jest.spyOn(mockAxios, 'post');
    store = new Vuex.Store({
      modules: {
        users,
      },
    });
    wp = shallowMount(UserRegister, { store, localVue });
  });

  it('html should render correctly', () => {
    expect(wp.html()).toMatchSnapshot(); // UI가 나중에 변경될까바 참고 : https://jestjs.io/docs/en/snapshot-testing
  });

  it('Component가 제대로 렌더 되는가', () => {
    expect(wp.exists()).toBe(true);
  });

  it('이메일, 비밀번호, 닉네임이 제대로 담기는가', () => {
    wp.find('.email').setValue(email);
    wp.find('.password').setValue(password);
    wp.find('.nickname').setValue(nickname);

    wp.find('.userRegister').trigger('click');

    wp.vm.$nextTick();

    expect(wp.vm.data.email).toBe(email);
    expect(wp.vm.data.password).toBe(password);
    expect(wp.vm.data.nickname).toBe(nickname);
  });

  const userCases = [
    ['', '', '', false],
    ['ke', null, '', false],
    ['ke', 'ke', 'ke', true],
  ];
  it.each(userCases)(
    'data의 fields는 null이 없어야한다.',
    (email, password, nickname, expected) => {
      wp.setData({
        data: {
          email: email,
          password: password,
          nickname: nickname,
        },
      });
      let res = wp.vm.validateData();
      expect(res).toEqual(expected);
    }
  );

  it('비밀번호가 서로 다르면 메세지를 띄워줘야한다', async () => {
    // NOTE v-if가 제대로 테스트가 안됬음. 해결책 https://github.com/vuejs/vue-test-utils/issues/1384
    expect(wp.find('.error__password--input').exists()).toBe(false);

    let wrongPwd = '12345';
    wp.find('.password').setValue(password);
    wp.find('.password_re').setValue(wrongPwd);
    await wp.vm.$nextTick();

    expect(wp.find('.error__password--input').exists()).toBe(true);
    let res = wp.find('.error__password--input').text();
    expect(res).toBe('패스워드가 맞지 않습니다.');
  });

  it('비밀번호 체크는 실시간으로 되야한다', () => {
    let wrongPwd = '12345';
    wp.find('.password').setValue(password);
    wp.find('.password_re').setValue(wrongPwd);
    expect(wp.vm.isPasswordSame).toBe(false);

    wp.find('.password').setValue(password);
    wp.find('.password_re').setValue(password);
    expect(wp.vm.isPasswordSame).toBe(true);
  });

  it('이메일 중복 검사 성공 / 실패 안내 문구 ', async () => {
    let getters = {
      apiStatus: () => {
        return {
          code: 200,
          msg: '',
        };
      },
    };
    users.getters = getters;
    store = new Vuex.Store({
      modules: {
        users,
      },
    });
    wp = shallowMount(UserRegister, { store, localVue });
    await expect(wp.find('.error__email--duplicated').exists()).toBe(false);
  });

  it('data를 담고 버튼을 클릭하면 data가 전송된다', async () => {
    wp.find('.email').setValue(email);
    wp.find('.password').setValue(password);
    wp.find('.nickname').setValue(nickname);

    wp.find('.userRegister').trigger('click');

    let data = {
      email: email,
      password: password,
      nickname: nickname,
    };
    await flushPromises();

    let url = getUrlFromSpy(spyPost);
    let body = getDataFromSpy(spyPost);

    expect(url).toBe('/api/users/v1');
    expect(body).toEqual({ data: data });
  });
});
