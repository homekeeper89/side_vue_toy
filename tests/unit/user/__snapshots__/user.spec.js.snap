// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Register 와 관련된 테스트 html should render correctly 1`] = `
<div class="UserRegister">
  <div><span>email : </span> <input type="text" class="email"></div>
  <!---->
  <div><span>password : </span> <input type="text" class="password"> <span>password_again : </span> <input type="text" class="password_re"></div>
  <!---->
  <div><input type="text" class="nickname"></div>
  <div><button class="userRegister">등록</button></div>
</div>
`;
